exchanges/subscription/subscription.go

```diff
@@ -1,92 +1,140 @@
-// If Pair becomes a pointer, this method is redundant and should be removed
-func (s *Subscription) MarshalJSON() ([]byte, error) {
-	// None of the usual type embedding tricks seem to work for not emitting an nil Pair
-	// The embedded type's Pair always fills the empty value
-	type MaybePair struct {
-		Enabled       bool                   `json:"enabled"`
-		Channel       string                 `json:"channel,omitempty"`
-		Asset         asset.Item             `json:"asset,omitempty"`
-		Params        map[string]interface{} `json:"params,omitempty"`
-		Interval      kline.Interval         `json:"interval,omitempty"`
-		Levels        int                    `json:"levels,omitempty"`
-		Authenticated bool                   `json:"authenticated,omitempty"`
-		Pair          *currency.Pair         `json:"pair,omitempty"`
-	}
-
-	k := MaybePair{s.Enabled, s.Channel, s.Asset, s.Params, s.Interval, s.Levels, s.Authenticated, nil}
-	if s.Pair != currency.EMPTYPAIR {
-		k.Pair = &s.Pair
-	}
-
-	return json.Marshal(k)
+// MatchableKey interface should be implemented by Key types which want a more complex matching than a simple key equality check
+type MatchableKey interface {
+	Match(any) bool
 }
 
 // String implements the Stringer interface for Subscription, giving a human representation of the subscription
 func (s *Subscription) String() string {
-	return fmt.Sprintf("%s %s %s", s.Channel, s.Asset, s.Pair)
+	p := s.Pairs.Format(currency.PairFormat{Uppercase: true, Delimiter: "/"})
+	return fmt.Sprintf("%s %s %s", s.Channel, s.Asset, strings.Join(p.Strings(), ","))
 }
 
-// EnsureKeyed sets the default key on a channel if it doesn't have one
-// Returns key for convenience
+// State returns the subscription state
+func (s *Subscription) State() State {
+	s.m.RLock()
+	defer s.m.RUnlock()
+	return s.state
+}
+
+// SetState sets the subscription state
+// Errors if already in that state or the new state is not valid
+func (s *Subscription) SetState(state State) error {
+	s.m.Lock()
+	defer s.m.Unlock()
+	if state == s.state {
+		return ErrInStateAlready
+	}
+	if state > UnsubscribingState {
+		return ErrInvalidState
+	}
+	s.state = state
+	return nil
+}
+
+// EnsureKeyed returns the subscription key
+// If no key exists then a pointer to the subscription itself will be used, since Subscriptions implement MatchableKey
 func (s *Subscription) EnsureKeyed() any {
 	if s.Key == nil {
-		s.Key = DefaultKey{
-			Channel: s.Channel,
-			Asset:   s.Asset,
-			Pair:    s.Pair,
-		}
+		s.Key = s
 	}
 	return s.Key
 }
+
+// Match returns if the two keys match Channels, Assets, Pairs, Interval and Levels:
+// Key Pairs comparison:
+// 1) Empty pairs then only Subscriptions without pairs match
+// 2) >=1 pairs then Subscriptions which contain all the pairs match
+// Such that a subscription for all enabled pairs will be matched when searching for any one pair
+func (s *Subscription) Match(key any) bool {
```

Comment:
This function does not seem to be used at all in GCT and can be removed. In future if used you can change from `any` to specific `*Subscription` type for key

Reply:
It's definitely used.
store.match calls it at line 147.
Note the documentation above Get and get; *Subscription implements MatchableKey. This *is* that implementation. A subscriptions default key is a pointer to itself.

Reply:
I am super confused, which exchange uses this? I just want to test it out while running GCT. I can delete this entire function, compile and build and it works fine without it.

Reply:
Any exchange that tries to Get() a subscription without giving it a bespoke key should use this.
The first of these will be Kraken.

Also don't get how we could change the type from any and retain cross-key compatibility.
Or non-pointer compatibility.

This is a stacked PR; I don't want to do a huge load of work specific to Kraken *whilst* introducing a new facility.
So I make the design, and then PR that as a contained and working solution, and then use it in Kraken.
If I don't do it that way, we get never ending PRs and have to disentangle the Subscription code when I have to merge/rebase it.

Reply:
Note: To be specific, I'm expecting around 80% of the exchanges will rely on the default key, since they don't otherwise key their subs.

Reply:
So above you said it was definitely used but now its not definitely used it's a a contained and working solution? I will defer this to @gloriousCode. I am not that smart man, if I can't test it in running GCT and slapping down some print lines I got no idea how this is working in a holistic sense. I generally don't like merging dead code, anymore, I used to do it all the time; for future possible things. I am obviously not getting something but I will go through it and do some testing soon instead of eyeballing and making a silly of myself which I do constantly.

Reply:
Sorry, I was no where near clear enough. Let me try again, and stick to essential information.

`Match` is used now.
Being able to delete it doesn't mean it's not used.
**Put a panic or exit in there and it'll be triggered.**
`Get()` calls it, and Get is used a lot.

Reply:
I can't comment on how things were ~4 days ago, but I'll just detail things. [It definitely does get hit ](https://github.com/thrasher-corp/gocryptotrader/assets/9261323/a6fef05d-99ca-4657-bd7d-5b41f7d3fe4a)

A `Subscription` implements `Match` which is defined in `MatchableKey`. There is the function `func (s *Store) get(key any) *Subscription` where `MatchableKey` is used to check if a `Subscription` matches one in the `Store`.
It allows you to define custom subscription keys and then to define how that custom key can be matched in the `Store` to get a Subscription back.

The reason why you can delete the function is that the `get` function has a `default` case in the event of key not implementing `MatchableKey` the Key becomes the `Subscription` itself. The `HobbitKey` test type is a little confusing overall, but demonstrates how you can store a `Subscription` with a different Key than checking for Exchange Asset Pair.

My guess is, Kraken will keep track of its own type of Key type, build that Key in `wsHandleData` then use that to call `Get` to get Subscription details. @gbjk I would appreciate seeing how it will properly look (versus HobbitKey), since I'm guessing you have a different branch with that there.

Reply:
@gloriousCode Spot on, and thank you 🙏 

I've improved the test case with 3149dbab - HobbitKey was too confusing and AllPairsKey actually makes sense.

Kraken will make use of map lookup, and if it didn't I probably would have considered removing the map in favour of a list.
What's unique about Kraken is that it uses a different native type for Subscribing and Subscribed keys, so it can re-find the subscription later.
I'll have that ready to see later this week.

Reply:
Completely rewritten this to make it much simpler.
There's a few use-cases here from the Kraken branch still, but very minor.